### PR TITLE
fix: bound implementation-session review loop with a code-enforced 2-pass cap

### PR DESCRIPTION
**Pull request by ivanviragine**
Closes #384

## Complexity
complex

## Context / Problem

During an implementation session an agent can get stuck in an infinite review
loop: `wade review implementation` reports findings, the agent fixes them and
commits, then `wade implementation-session done` refuses to complete — telling
the agent to review again — and the cycle repeats without end.

The "run review at most 2 times" rule that is supposed to stop this **exists
today only as prose** in `templates/skills/_partials/review-implementation-closing-step.md:22-28`
(and mirrored in project knowledge `cedeaad0`). Nothing in Python counts review
passes or bounds the loop.

Two concrete mechanics produce the stuck state:

1. **Commit-after-review (by design, but unbounded).** The review writes a
   SHA-keyed marker `.wade/reviewed@<sha>` (`_mark_reviewed`,
   `src/wade/services/review_delegation_service.py:29-44`). The completion gate
   `_gate_review_ran` (`src/wade/services/implementation_service/done.py:375-400`)
   refuses unless a marker exists for the current branch tip
   (`head_sha` = `pre_sync_head`, captured at `done.py:244-248`). Any commit the
   agent makes *after* its last review moves the tip SHA and invalidates the
   marker → "new commits require a fresh review" (`done.py:391-399`) → the agent
   re-reviews, commits again, and loops. Re-reviewing new commits is intended;
   the missing piece is a cap on how many times it can repeat.

2. **Headless review silently drops the marker.** `_mark_reviewed()` is called
   only when `result.success` is true (`review_delegation_service.py:300-301`).
   In headless mode — this repo's config (`.wade.yml:29` `mode: headless`) —
   `success` is just the subprocess exit code
   (`src/wade/services/delegation_service.py:130-144`). A ~300s timeout on a
   large diff exits non-zero (confirmed by knowledge `271cf18f`), so the review
   *ran and produced findings* but **no marker is written**, and `done` then
   refuses forever even though the human/agent saw a completed review.

`--skip-review` already exists on `done`
(`src/wade/cli/implementation_session.py:114-116`, consumed at `done.py:389`) as
a full manual bypass, but relying on the agent to remember a flag is fragile and
skips review entirely rather than bounding it.

## Proposed Solution

Move the 2-pass cap out of skill prose and into the completion gate, so the loop
is bounded in code while re-review of new commits is preserved.

1. **Review-pass count via a per-SHA `review-pass@<sha>` marker family.** Do not
   invent an integer counter file — reuse the existing `markers.py` primitives.
   On every **delegation-backed** review run, write a `review-pass@<HEAD>` marker
   (a distinct name from `reviewed`, so `write_marker`'s clear-prior-same-name
   cleanup does not touch it). The "pass count" is then simply *the number of
   distinct SHAs that carry a `review-pass@*` marker* — count matching filenames,
   defaulting to `0` if the directory listing fails. This design:
   - is **independent of `result.success`** (a headless timeout still writes the
     marker), sidestepping the marker-drop mechanic without certifying an
     unfinished review;
   - is naturally **fail-safe** — there is no numeric content to corrupt, and a
     listdir failure yields `0`, never a false "cap reached";
   - is naturally **idempotent per SHA** — this closes a retry loophole: a plain
     counter would count *retries on the same unchanged commit* (e.g. re-running a
     timed-out review twice on the same HEAD would exhaust a cap of 2 with zero
     real review→fix cycles). Keying to distinct SHAs makes same-commit retries
     free and only spends budget on genuine review→fix→new-commit cycles, which is
     what the cap is meant to bound;
   - requires **no reset logic and no `bootstrap_worktree` change** — markers
     persist naturally across `wade implement` re-runs; just ensure bootstrap does
     not clear `review-pass@*`.

2. **Bounded completion gate.** Rewrite `_gate_review_ran`
   (`src/wade/services/implementation_service/done.py:375-400`) to:
   - **Fast path (unchanged):** if `marker_present(worktree_root, "reviewed",
     head_sha)` → pass. A review for the exact current commit still means done.
   - **Bounded escape (new, implementation session only):** else if review-pass
     count ≥ `max_review_passes` (default 2) → **pass**, printing a prominent
     notice that the 2-pass safety limit was hit, the latest commit(s) were not
     re-reviewed, and completion is being allowed anyway (mention `--skip-review`
     for an explicit full bypass).
   - **Refuse (improved):** else → refuse as today, but the message must state
     "review pass X of N used" and point the agent at both re-running review and
     `wade implementation-session done --skip-review`.
   - **Fail-closed count:** the pass-count is derived from counting `review-pass@*`
     markers; a listing failure yields `0` (never a false "cap reached"),
     consistent with `markers.py`'s best-effort "fail toward re-gating" convention.

3. **Make the cap configurable, bounded at both boundaries.** Add
   `max_review_passes` (default `2`) to `DoneConfig`
   (`src/wade/models/config.py:159-182`) with a **Pydantic constraint**
   (`PositiveInt` or `Field(gt=0)`) so a bad `.wade.yml` value
   (`done.max_review_passes: 0` / `-1`) fails loudly *at load*, not just at
   `wade check`. Parse it in `src/wade/config/loader.py:267-279` with its **own**
   branch — do **not** reuse `_done_flag` (`loader.py:269-273`), which assumes a
   bool default; read `done_raw.get("max_review_passes", 2)` and normalize an
   explicit `null` to `2`. The `wade check` validator (item 7) adds the second,
   lint-boundary guard.

4. **No reset logic required.** Because the count is the number of distinct
   `review-pass@<sha>` markers (item 1), it persists naturally across
   `wade implement` re-runs and needs no `bootstrap_worktree` involvement — which
   is important, since `bootstrap_worktree` runs on *every* invocation, including
   the idempotent worktree-reuse path (`implementation_service/core.py:433` →
   `bootstrap_worktree` at `:492`) and the review re-bootstrap
   (`review_service.py:683`); anything it touched would have to be idempotent
   anyway. Add a test that the count **survives** a second `wade implement`
   invocation on the same worktree.

5. **Correctly scope the cap to the implementation session.** `_gate_review_ran`
   is **shared** — `_run_completion_gates` calls it on *both* the
   `review-pr-comments` path and the implementation path
   (`done.py:321-331`), and `done()` is one service for both session types
   (knowledge `851bb6ec`). Adding counter/cap logic directly inside
   `_gate_review_ran` would therefore change `review-pr-comments-session done`
   too. **Fix:** thread `session_type` (or an explicit bool) into
   `_gate_review_ran` and apply the counter/cap branch only on the implementation
   path; leave `review-pr-comments` on the existing fast-path-or-refuse behavior.

6. **Do not consume a cap slot on the no-diff early return.** `review_implementation`
   already calls `_mark_reviewed()` on its "no changes to review" path
   (`review_delegation_service.py:274`) without going through delegation. That
   path writes the exact-SHA marker (which already satisfies the gate fast path),
   so it must **not** write a `review-pass@<sha>` marker — only a
   delegation-backed review run consumes a cap slot.

7. **Update the `wade check` done validator.** `_validate_done_section`
   (`src/wade/services/check_service.py:657-670`) hard-codes bool-only validation
   (`if not isinstance(value, bool): "must be true or false"`). `_VALID_DONE_KEYS`
   (`check_service.py:266`) auto-picks up the new field so it won't be flagged as
   "unsupported", but any int value (`done.max_review_passes: 3`) will fail
   `wade check`. Special-case `max_review_passes`: accept a positive int, reject
   bool / zero / negative with a clear message.

8. **Sync the skill prose.** Update
   `templates/skills/_partials/review-implementation-closing-step.md:22-28` so it
   states the 2-pass cap is now enforced in code, and tell the agent that a
   genuine unavoidable loop can be broken with
   `wade implementation-session done --skip-review`. Keep the existing exit-code
   guidance (0/1/2) intact. (No renumbering of `## Your role` steps is required —
   only body text of the partial changes.)

**Non-goals:**
- **`review-pr-comments-session` is intentionally out of scope.** The same two
  root-cause mechanics apply to `_gate_review_ran` on that path, but the bug
  report is scoped to implementation sessions and that gate is shared. This change
  caps the implementation path only; `review-pr-comments-session done` can still
  loop and is tracked separately.
- Not changing the `reviewed@<sha>` marker semantics, not altering headless
  success/exit-code handling in `delegation_service`, and not touching the
  `done@<sha>` marker used by the Stop hook / pre-push backstop.

## Tasks
- [ ] Add a helper (in/near `src/wade/utils/markers.py`) that counts distinct
      `review-pass@*` markers, reusing existing marker primitives; a directory
      listing failure returns `0` (fail-safe). Use a name distinct from `reviewed`
      so `write_marker`'s clear-prior-same-name cleanup does not remove it.
- [ ] Write a `review-pass@<HEAD>` marker in `review_implementation`
      (`src/wade/services/review_delegation_service.py`) whenever the **delegation**
      runs, independent of `result.success`. Do **not** write it on the "no
      changes to review" early-return path (`:274`) — it already writes the
      exact-SHA `reviewed` marker and must not consume a cap slot. (Per-SHA writes
      are idempotent, so same-commit retries do not inflate the count.)
- [ ] Add `max_review_passes` (default `2`) to `DoneConfig`
      (`src/wade/models/config.py`) with a Pydantic bound (`PositiveInt` /
      `Field(gt=0)`), and parse it in `src/wade/config/loader.py` with its **own**
      null-handling branch (not `_done_flag`, which is bool-only).
- [ ] Special-case `max_review_passes` in `_validate_done_section`
      (`src/wade/services/check_service.py:657-670`): accept a positive int, reject
      bool / zero / negative, so `wade check` does not fail on
      `done.max_review_passes: 3`.
- [ ] Thread `session_type` (or a bool) into `_gate_review_ran`
      (`done.py:375-400`, both call sites at `done.py:321-331`) and apply the
      fast-path / bounded-escape / improved-refuse logic **only on the
      implementation path**; keep `review-pr-comments` on the existing
      fast-path-or-refuse behavior. Preserve the `skip_review`, `require_review`,
      and `review_implementation.enabled` short-circuits.
- [ ] Confirm `bootstrap_worktree` does **not** clear `review-pass@*` markers
      (it re-runs on the idempotent `wade implement` path, `core.py:433/492`, and
      review re-bootstrap, `review_service.py:683`); the marker family needs no
      dedicated reset logic.
- [ ] Update `templates/skills/_partials/review-implementation-closing-step.md`
      prose to reflect the code-enforced cap and mention `--skip-review` as the
      loop escape.
- [ ] Unit tests: a `review-pass@<sha>` marker is written on a delegation-backed
      review run incl. the headless-timeout (`result.success=False`) path; no marker
      on the no-diff early return; **two reviews on the same HEAD sha count as one
      pass** (idempotent per-SHA); count is `0` when the marker-dir listing fails
      (fail-safe); **count survives a second `wade implement` invocation on the same
      worktree**; gate passes on exact-SHA `reviewed` marker; gate passes after N
      passes with the notice; gate refuses before N with the "pass X of N" message;
      `max_review_passes` config override honored and a `<= 0` value is rejected at
      load (Pydantic) and by `wade check`; **`review-pr-comments-session done` path
      unaffected by the cap**.
- [ ] Deterministic E2E contract test: review → commit → `done` refuses → review
      again → commit → `done` **succeeds** on the 2nd pass instead of looping.
- [ ] Run `./scripts/check-all.sh` (tests + lint + strict mypy) and fix any
      findings.
- [ ] Update `README.md` if the user-facing `done`/`--skip-review` behavior is
      documented there; update `AGENTS.md`/`docs/dev/` only if the gate contract
      changed materially. Consider correcting/refreshing knowledge `851bb6ec` if
      the `_gate_review_ran` sharing detail proves worth recording.

## Acceptance Criteria
- [ ] A commit made after the last review no longer produces an unbounded loop:
      after at most `max_review_passes` (default 2) runs of
      `wade review implementation`, `wade implementation-session done` completes
      instead of refusing.
- [ ] When the cap is reached, `done` prints a clear notice that the 2-pass limit
      was hit, the latest commit(s) were not re-reviewed, and that `--skip-review`
      is the explicit full-bypass.
- [ ] Before the cap, the refuse message states the current pass number out of the
      limit and points the agent at re-running review and at `--skip-review`.
- [ ] A headless review that exits non-zero (e.g. timeout) still counts toward the
      cap, so it cannot cause an infinite loop even though no `reviewed@<sha>`
      marker is written.
- [ ] The existing exact-SHA fast path is unchanged: a review for the current
      commit lets `done` proceed on the first pass.
- [ ] `done.max_review_passes` in `.wade.yml` overrides the default; omitting it
      keeps the default of 2. Setting `done.require_review: false` or
      `ai.review_implementation.enabled: false` still bypasses the gate as today.
- [ ] A `done.max_review_passes` of `0` / negative is rejected both at config load
      (Pydantic) and by `wade check`; a positive int is accepted.
- [ ] Re-running `wade review implementation` on the **same** HEAD sha does not
      advance the pass count (per-SHA markers are idempotent); only a new commit's
      review consumes a slot.
- [ ] The pass count (number of distinct `review-pass@<sha>` markers) **survives** a
      second `wade implement` invocation (and a review re-bootstrap) on the same
      worktree rather than resetting to 0.
- [ ] A failure to list the marker directory yields pass-count 0 (fail-safe), never
      a false "cap already reached".
- [ ] `review-pr-comments-session done` behavior is unchanged (the cap applies to
      the implementation path only).
- [ ] The `review-implementation-closing-step.md` skill partial documents the
      code-enforced cap and the `--skip-review` escape.
- [ ] `./scripts/check-all.sh` passes (tests, lint, strict mypy), including the
      new unit tests and the E2E loop-escape contract test.

## Summary by CodeRabbit

- **New Features**
  - Added a configurable implementation review-pass limit, defaulting to two.
  - Review progress persists across reruns, with safe handling for interrupted reviews.
  - Review-comment sessions remain uncapped.
  - Temporary-directory writes are allowed during regular execution, but not plan mode.
  - Increased the default delegation timeout to 600 seconds.

- **Bug Fixes**
  - Prevented indefinite review-and-fix cycles.
  - Invalid review-pass limits are rejected.

- **Documentation**
  - Updated configuration, review guidance, timeout handling, and recovery instructions.

- **Tests**
  - Added coverage for limits, persistence, validation, retries, and edge cases.

## What was done

Bounds the implementation-session review→fix→re-review loop with a **code-enforced
2-pass cap**. Previously the "run review at most 2 times" rule existed only as
skill prose, so an agent could loop forever: `wade review implementation` reports
findings → agent commits a fix → `wade implementation-session done` refuses (the
new commit invalidates the SHA-keyed `reviewed@<sha>` marker) → repeat. A headless
review that times out (exit non-zero) made it worse — it produced findings but
wrote no marker, so `done` refused indefinitely. The cap now lives in the
completion gate: after `done.max_review_passes` (default 2) genuinely-new commits
have been reviewed, `done` completes with a notice instead of looping, while
re-review of new commits is still preserved.

## Changes

- **`utils/markers.py`** — new `review-pass@<sha>` marker family (distinct name
  from `reviewed`, so `write_marker`'s clear-prior-same-name cleanup never touches
  it). `record_review_pass()` writes it **without** clearing prior markers;
  `count_review_passes()` returns the number of distinct `review-pass@*` markers,
  fail-safe to `0` on any directory-listing failure (never a false "cap reached").
- **`services/review_delegation_service.py`** — `review_implementation` records a
  `review-pass@<HEAD>` marker on every delegation-backed run, **independent of
  `result.success`** (a headless timeout still counts), so the loop can't be
  infinite. Not recorded on the no-diff early return (which already writes the
  exact-SHA `reviewed` marker and must not spend a cap slot). Per-SHA writes are
  idempotent, so re-reviewing the same HEAD does not advance the count.
- **`services/implementation_service/done.py`** — `_gate_review_ran` now takes
  `session_type`. On the implementation path: exact-SHA fast path unchanged →
  else bounded escape at `>= max_review_passes` (prominent notice, mentions
  `--skip-review`) → else refuse with "review pass X of N" and pointers to
  re-running review and `--skip-review`. The shared `review-pr-comments` path keeps
  the existing unbounded fast-path-or-refuse behavior (cap is impl-only, #384 scope).
- **`models/config.py` + `config/loader.py`** — `DoneConfig.max_review_passes:
  StrictInt = Field(default=2, gt=0)` — a `0`/negative **or non-int** value
  (`true`/`"2"`/`2.0`) fails loudly at load (strict, not coerced), parsed with its
  own null-handling branch (not the bool-only `_done_flag`).
- **`services/check_service.py`** — `wade check` special-cases `max_review_passes`
  (accepts a positive int; rejects bool / zero / negative).
- **Skill + docs** — `_partials/review-implementation-closing-step.md` now states
  the cap is code-enforced and points at `--skip-review` as the loop escape (kept
  under the session-start char budget). README completion-gate table and
  `docs/dev/architecture.md` review-ran gate description updated.

## Key technical decisions

- **Count distinct SHAs, not an integer counter.** Reusing the existing marker
  primitives (rather than a counter file) makes the design fail-safe (nothing to
  corrupt; a listdir failure yields 0), idempotent per SHA (same-commit retries are
  free — only real review→fix→new-commit cycles spend a slot), and reset-free
  (markers persist across `wade implement` re-runs, so no `bootstrap_worktree`
  change is needed).
- **Scoped to the implementation session.** `_gate_review_ran` is shared by both
  `done` session types (knowledge 851bb6ec), so the cap branch keys on
  `session_type == "implementation"`; `review-pr-comments-session done` is
  intentionally out of scope and unchanged.

## Testing

- Unit: marker write/count/idempotency/fail-safe; pass recorded on the
  headless-timeout (`success=False`) path and **not** on the no-diff return; gate
  bounded-escape + improved-refuse messages + impl-only scoping; `max_review_passes`
  config load (default/override/null/`<=0` rejected) and `wade check` validation.
- E2E contract: review→commit→`done` **refuses** ("review pass 1 of 2"), then
  review→commit→`done` **succeeds** at the cap ("safety limit reached"); and the
  review-pass count **survives a second `wade implement`** on the same worktree.
- `./scripts/check-all.sh` green — 2796 unit/integration tests pass, ruff clean,
  strict mypy clean; deterministic E2E suite passes (69/69).

## Notes for reviewers

- The exact-SHA fast path and the `--skip-review` / `done.require_review: false` /
  `ai.review_implementation.enabled: false` short-circuits are all preserved.
- The `implement` session-start skill payload is budget-pinned at 8000 chars; the
  closing-step partial's re-review guidance was kept net-neutral (47 chars of
  headroom, unchanged by this review round), so no budget bump was needed.

## Review round (PR #385 — CodeRabbit)

6 threads across 5 files; 5 fixed, 1 declined with a posted reply. All resolved.

**Fixed**

- **Strict `done.max_review_passes` at load** (`config.py`, `loader.py`): the
  `PositiveInt` bound coerced `true`→1, `"2"`→2, `2.0`→2, so normal load silently
  accepted values `wade check` rejects. Now `StrictInt` + `gt=0` rejects them at
  `DoneConfig` construction (→ `ConfigError`). Added loader tests for `true` /
  `"2"` / `2.0`.
- **`count_review_passes` fails closed** (`markers.py`): the path-based `iterdir()`
  fallback followed a symlinked `.wade`, letting a crafted worktree plant
  `review-pass@*` entries to satisfy the cap. Removed the fallback — returns `0`
  when no descriptor-based directory read is available.
- **Re-review after committing any finding** (`review-implementation-closing-step.md`):
  "Minor findings: fix and proceed" was wrong — committing a fix stales the
  exact-sha `reviewed` marker and `done` refuses at pass 1 of 2. Now instructs
  commit + re-review; major-finding flow, cap, and escape hatch preserved.
- **Exact-sha precedence under an exhausted cap** (`test_done_gates.py`): the test
  now records the cap's worth of `review-pass` markers before `reviewed@head` and
  asserts no safety-limit notice, proving the fast path precedes the cap.

**Declined (reply posted on thread)**

- **"Scope review-pass markers per session / clear at fresh `wade implement`
  start."** The count is deliberately worktree-persistent (pinned by
  `test_review_pass_count_survives_second_implement`). Clearing on a fresh
  `implement` would make the cap trivially resettable — re-running `implement`
  would zero it and reinstate the unbounded loop #384 exists to stop. A worktree
  is bound 1:1 to a branch/issue (re-running `implement` is the same task; a
  recreated worktree starts with a fresh `.wade/`), so there is no cross-branch
  counter to scope; the fail-closed fix above covers the symlink-tampering angle.

## Follow-up: friendlier headless defaults (same PR)

Two default changes so headless review/deps stop tripping over infra limits
(the trigger: the review subprocess timed out at 300s while addressing this PR):

- **Default headless subprocess timeout 300s → 600s** (`DelegationRequest.timeout`,
  `models/delegation.py`). Applies to every wade project; per-command
  `ai.<cmd>.timeout` still overrides. A 300s budget killed a high-effort review
  over a large diff mid-run (an infra timeout, not a review result), which wrote
  no `reviewed` marker and stalled `done`.
- **Shell write guard allows system temp dirs** (`hooks/policies.py`). `/tmp` and
  `$TMPDIR` (resolved; trailing separator so `/tmpfoo` can't match `/tmp`) join
  `_ALWAYS_ALLOWED_PATH_PREFIXES`, so agents/wade can stage scratch files there.
  **Scoped to `shell_containment`** — the file-path guard `worktree_containment`
  stays worktree-only (it backs `plan_artifact_only`, where an artifact-named temp
  path such as `/tmp/.claude/plans/x.md` must not escape), and plan mode still
  denies temp writes as non-artifacts.

Docs (`architecture.md`) and the closing-step budget note updated; the
`shell_containment` tests moved their "outside" archetype to `/etc` and added
temp-allowed cases.

## Review round 2 (PR #385 — CodeRabbit)

2 threads across 2 files; both fixed and resolved.

**Fixed**

- **Describe the review-pass cap as configurable, not a hardcoded 2**
  (`markers.py`, `test_done_gates.py`, `review-implementation-closing-step.md`,
  `architecture.md`): `done.max_review_passes` is configurable (default 2), so
  the docstrings, test-class docstring, closing-step template header, and
  architecture gate description now describe the *configured* cap
  (`done.max_review_passes`, default 2) rather than a fixed "2-pass cap". For
  consistency I also dropped the hardcoded "2-pass" phrasing from the remaining
  internal comments (`done.py`, `review_delegation_service.py`) that the reviewer
  didn't list but describe the same cap. The plan-session review loop
  (`review-plan-step.md`, "run at most 2 times") is a separate, genuinely fixed
  constraint and was intentionally left unchanged.
- **Quote the temp-dir redirect target in the shell-containment test**
  (`test_policies.py`): `tempfile.gettempdir()` can return a path with
  whitespace or shell metacharacters, so the unquoted redirect could target a
  different path and pass for the wrong reason. Added `import shlex` and wrapped
  the target in `shlex.quote()`.

**Follow-up — surface the review-pass budget (from the closing `wade review
implementation` self-review)**

The self-review flagged that the cap is enforced silently at `done` time and its
`done`-time notice implied an unbounded loop even when the cap was simply reached
by reviewing distinct commits, while the "run at most N times" rule lived only in
buried skill prose. Rather than re-engineer the count (a reset-on-clean-review
scheme can't be built on the current markers without reintroducing the unbounded
loop, and round 1 already declined a resettable count as gameable), the fix makes
the existing count **visible**:

- `wade review implementation` now prints the running budget after each pass
  (`"review pass N of M — K left"`, or a cap-reached warning) — the count comes
  from the command itself, not buried prose. (`review_delegation_service.py`:
  `_record_review_pass` returns the count; new `_announce_review_pass_budget`.)
- The `done`-time notice now states the honest semantics (N commits reviewed on
  this worktree; current commit not re-reviewed) instead of always implying a
  loop, and points at both `done.max_review_passes` and `--skip-review`.
  (`done.py`.)
- Closing-step skill and `architecture.md` point at the command output instead of
  a hardcoded "run at most N times" rule.
- The count stays cumulative and configurable via `done.max_review_passes`
  (already `StrictInt`, `gt=0`, default 2); counting logic is unchanged. New tests
  in `test_reviewed_marker.py` cover the budget message (remaining/plural/reached)
  and the command wiring.

### Session 1

| Metric | Value |
| --- | --- |
| Tool | `claude` |
| Model | `claude-opus-4.8` |
| Total tokens | **231,400** |
| Input tokens | **230,100** |
| Output tokens | **1,300** |
| Cached tokens | **0** |

### Session 1

| Metric | Value |
| --- | --- |
| Tool | `claude` |
| Model | `claude-opus-4.8` |
| Total tokens | **314,168** |
| Input tokens | **313,200** |
| Output tokens | **968** |

### Session 2

| Metric | Value |
| --- | --- |
| Tool | `claude` |
| Model | `claude-opus-4.8` |
| Total tokens | **209,602** |
| Input tokens | **209,600** |
| Output tokens | **2** |
| Cached tokens | **0** |

## AI Sessions

| Phase | Tool | Session |
| --- | --- | --- |
| Implement | `claude` | `1b9dcc11-182c-49d9-8bbd-0f816943ca3e` |
| Review | `claude` | `a59ed06b-c68a-4f7c-b29f-821c887bdd8e` |
| Review | `claude` | `434ee082-6462-4652-8d08-2d68cffaeabd` |

### CODE REVIEW

File: src/wade/utils/markers.py [hunk 1]
```diff
@@ -226,6 +226,66 @@ def clear_markers(worktree_root: Path, name: str) -> None:
...
+            return 0
+        finally:
+            if dir_fd is not None:
+                os.close(dir_fd)
+    else:
+        try:
+            entries = [p.name for p in wade_dir.iterdir()]
+        except OSError:
+            return 0
+    return sum(1 for e in entries if e.startswith(prefix) and not e.endswith(".tmp"))
```
[the remark this answers is not in the document]
Comment (ivanviragine) [reply src/wade/utils/markers.py:290]: Declining this one after verifying against the current design.

The review-pass count is **deliberately** worktree-persistent, pinned by an E2E test (`test_review_pass_count_survives_second_implement`). Clearing the `review-pass@*` family at a fresh `wade implement` start would make the cap trivially resettable — an agent stuck in the loop could just re-run `implement` to zero the counter and loop forever, defeating #384's entire purpose.

On the stale-cross-session concern: a WADE worktree is bound 1:1 to a branch/issue, so re-running `implement` on it *is* the same logical task (preserving the count is correct), and a removed+recreated worktree starts with a fresh `.wade/`. There is no worktree-wide counter shared across unrelated branches to scope. The sibling fail-closed fix (returning 0 when descriptor-based reads are unavailable) covers the symlink-tampering angle raised in the adjacent thread.
